cg/meta/archive/ddn.py

```diff
@@ -0,0 +1,141 @@
+from cg.io.controller import APIRequest, ReadStream
+from cg.models.cg_config import DDNConfig
+
+
+class DDNApi:
+    """Class for archiving and retrieving folders via DDN."""
+
+    def __init__(self, config: DDNConfig):
+        self.database_name: str = config.database_name
+        self.user: str = config.user
+        self.password: str = config.password
+        self.url: str = config.url
+        self.archive_repository = config.archive_repository
+        self.local_storage = config.local_storage
+        self.auth_token: str
+        self.refresh_token: str
+        self.token_expiration: datetime
+        self._set_auth_tokens()
+        self.headers: Dict[str, str] = {
+            "Content-Type": "application/json",
+            "accept": "application/json",
+        }
+        self.ostype: str = "Unix/MacOS"
+
+    def _set_auth_tokens(self) -> None:
+        """Retrieves and sets auth and refresh token from the REST-API."""
+        response: Response = APIRequest.api_request_from_content(
+            api_method=APIMethods.POST,
+            url=urljoin(base=self.url, url="auth/token"),
+            headers={"Content-Type": "application/json", "accept": "application/json"},
+            json={
+                "dbName": self.database_name,
+                "name": self.user,
+                "password": self.password,
+                "superUser": False,
+            },
+        )
+        response_content: dict = ReadStream.get_content_from_stream(
+            file_format=FileFormat.JSON, stream=response.content
+        )
+        self.refresh_token: str = response_content.get("refresh")
+        self.auth_token: str = response_content.get("access")
+        self.token_expiration: datetime = datetime.fromtimestamp(response_content.get("expire"))
+
+    def _refresh_auth_token(self) -> None:
+        """Updates the auth token via by providing the refresh token to the REST-API."""
+        response: Response = APIRequest.api_request_from_content(
+            api_method=APIMethods.POST,
+            url=urljoin(base=self.url, url="auth/token/refresh"),
+            headers=self.headers,
+            json={
+                "refresh": self.refresh_token,
+            },
+        )
+        response_content: dict = ReadStream.get_content_from_stream(
+            file_format=FileFormat.JSON, stream=response.content
+        )
+        self.auth_token: str = response_content.get("access")
+        self.token_expiration: datetime = datetime.fromtimestamp(response_content.get("expire"))
+
+    @property
+    def auth_header(self) -> Dict[str, str]:
+        """Returns an authorisation header based on the current auth token, or updates it if
+        needed."""
+        if datetime.now() > self.token_expiration:
+            self._refresh_auth_token()
+        return {"Authorization": f"Bearer {self.auth_token}"}
+
+    def archive_folders(self, sources_and_destinations: Dict[Path, Path]) -> bool:
```

Comment:
Not sure if the `Dict[Path, Path]` type is the clearest/best for this function, so feel free to pitch in with ideas